### ISSUE TITLE
[C] Error Detecting Functions Defined with Macro, Return Type and Function Name on Different Lines

### ISSUE THREAD

**Issue by alan33d** [sublimehq/Packages#433]
In previous builds, I can using Command + R on OS X (10.11) to properly highlight and scroll through Functions. The issue in Build 3114 is it will only pick out the return variable type (see example) if the return variable type and the function name are not on the same line.  Below example using C + UEFI.

3114 Issue:
In previous versions, Command + R will show you "foo" in the "@" function scroller. But now it shows just "EFI_STATUS". Just showing the return type is not useful. The editor was smarter in the previous versions where it can pick out "foo".

Can someone help fix this?

In UEFI, EfiBind.h
```c++
...
typedef uint32_t UINT32;
...
typedef UINT32 EFI_STATUS;
...
```
Then if I do the following, it doesn't work.

```c++
EFI_STATUS
foo (int var1) {
EFI_STATUS var2
...
return var2;
}

**Comment by wbond** [comment]
Here is the code sample I am working from:

```c++
typedef uint32_t UINT32;
typedef UINT32 EFI_STATUS;


EFI_STATUS
foo (int var1) {
	EFI_STATUS var2;
	return var2;
}
foo (int var1) {
	int var2;
	return var2;
}
```

The symbol list shows the following:

 - The `typedef` of `UINT32`
 - The `typedef` of `EFI_STATUS`
 - The function `foo`

The second foo doesn't appear in the symbol list since there isn't a good way to detect that it is a function definition instead of a function call since it is using the old C89 implicit return type. This has to do with the details of how the Sublime Text lexer is implemented. Throwing an `int` in front of the name would be the most compatible thing to do.

Previously certain typedefs and other data structures were not shown in the Goto Symbol list, but we improved that so that functionality such as Goto Definition would work properly.

**Comment by alan33d** [comment]
Sorry the second foo was a typo.  

My issue is Build 3114 is now picking up EFI_STATUS in the Goto Symbol list rather than foo.  In this case, I need a way to scroll through my functions and it's not picking up the function names.  Build 3103 worked OK.  Do you have a download link for that on OS X?  

Any chance the Goto Symbol List feature can be improved to pick up function names in any situation for C/C++?

**Comment by wbond** [comment]
From my testing it is picking up `foo` and `EFI_STATUS`. If it isn't working for other code, can you please paste the code here so I can look into it?

**Comment by alan33d** [comment]
Code and pictures attached to show issue in Build 3114 but not in Build 3103.
![screen shot 2016-05-26 at 2 15 08 pm](https://cloud.githubusercontent.com/assets/11778990/15590380/53dfbcc2-234c-11e6-8383-52d0e7392c93.png)
![screen shot 2016-05-26 at 2 14 41 pm](https://cloud.githubusercontent.com/assets/11778990/15590381/53f4bb5e-234c-11e6-94ee-40da256c2dd6.png)
[test.c.zip](https://github.com/sublimehq/Packages/files/285345/test.c.zip)

**Comment by OdatNurd** [comment]
It looks like this sometimes works and sometimes does not. For example:
![sublime2](https://cloud.githubusercontent.com/assets/13212876/15592675/7a486b80-235a-11e6-8f90-bcfc2f484f85.png)

The definition works across multiple lines, but only if the storage modifier is explicitly known and not a define or what have you; in that case it picks up the return type as a function declaration and the function name as a function call.

**Comment by alan33d** [comment]
This needs to be fixed to be smarter.  Again Build 3103 worked for obtaining function names in the Goto Symbols tool.  This is an extremely helpful feature that should work in all cases to detect function names.  It'll be ashamed if it only works in some cases.

**Comment by wbond** [comment]
See, there were many situations it didn't work in 3103, hence the rewrite. We added detection of all sorts of complex situations. It just so happens the format in this case isn't easy to detect.

The way you have a macro, return type and function name split over multiple lines makes it hard to detect properly. Most likely if we try to detect all uppercase at the end of a line without a ( as a macro/type will help in some situations. See some users put the parens on a separate line, so we have to use heuristics to find the function name.

If you put the return type on the same line as the function name, it will work currently.

I'll try to do some further tuning.

**Comment by alan33d** [comment]
We have millions of lines of code written in this style.  And most of it is in production.  It wouldn't be kosher for me to just change the format because I use Sublime Text 3 Build 3114.  

I will continue to use Build 3103 for now until the next release where this will hopefully be fixed.

Thank you for your help.

**Comment by wbond** [comment]
This should be resolved by 6a5e2d46d01df3c5da0d3cd181ea62fffdc7aa76
